### PR TITLE
Add notification_url signature checking method

### PR DESCRIPTION
**Pull request by RTLcoil**

### CODE REVIEW

File: cloudinary/api.py [hunk 1]
```diff
@@ -510,3 +511,42 @@ def __delete_resource_params(options, **params):
              **only(options, "keep_original", "next_cursor", "invalidate"))
     p.update(params)
     return p
+
+
+def verify_api_response_signature(public_id, version, signature):
```
Comment (const-cloudinary) [inline cloudinary/api.py]: I think these two should be moved out of api.py to a separate file, similar to PHP implementation. these are separate helper methods not that related to the API itself

File: test/test_api.py [hunk 2]
```diff
@@ -728,6 +734,61 @@ def test_update_access_control(self, mocker):
         self.assertIn("access_control", params)
         self.assertEqual(exp_acl, params["access_control"])
 
+    @unittest.skipUnless(cloudinary.config().api_secret, "requires api_key/api_secret")
+    def test_verify_api_response_signature(self):
```
Comment (const-cloudinary) [inline test/test_api.py]: move tests to separate file as well, and try to align input and expected values with PHP SDK.

File: cloudinary/utils.py [hunk 3]
```diff
@@ -1264,6 +1264,56 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def compute_hex_hash(s):
```
Comment (const-cloudinary) [inline cloudinary/utils.py]: When you add a helper method, please update all code with the new method, search for `hashlib.sha1` in the project and replace with this function (where applicable)

File: cloudinary/utils.py [hunk 4]
```diff
@@ -1264,6 +1264,56 @@ def wrapper(*args, **kwargs):
...
+    :param s: string to process
+
+    :return: HEX string
+    """
+    return hashlib.sha1(to_bytes(s)).hexdigest()
+
+
+def verify_api_response_signature(public_id, version, signature):
+    """
+    Validates API response signature against Cloudinary configuration.
```
Comment (const-cloudinary) [inline cloudinary/utils.py]: Please use doc strings from the reference implementation

File: cloudinary/utils.py [hunk 5]
```diff
@@ -1264,6 +1264,56 @@ def wrapper(*args, **kwargs):
...
+    parameters_to_sign = {'public_id': public_id,
+                          'version': version}
+    signed_parameters = api_sign_request(parameters_to_sign, cloudinary.config().api_secret)
+
+    return signature == signed_parameters
+
+
+def verify_notification_signature(body, timestamp, signature, valid_for=7200):
+    """
+    Validates notification signature against Cloudinary configuration
```
Comment (const-cloudinary) [inline cloudinary/utils.py]: Please use doc strings from the reference implementation

File: cloudinary/utils.py [hunk 6]
```diff
@@ -1264,6 +1264,56 @@ def wrapper(*args, **kwargs):
...
+
+    :param public_id: Public ID of resource
+    :param version: Version of resource
+    :param signature: Response signature
+
+    :return: Boolean result of the validation
+    """
+    parameters_to_sign = {'public_id': public_id,
+                          'version': version}
+    signed_parameters = api_sign_request(parameters_to_sign, cloudinary.config().api_secret)
```
Comment (const-cloudinary) [inline cloudinary/utils.py]: Please verify that api_secret is not empty and throw exception if it is. Add unit test.
:+1: for reusing `api_sign_request`

File: cloudinary/utils.py [hunk 7]
```diff
@@ -1264,6 +1264,56 @@ def wrapper(*args, **kwargs):
...
+
+    :return: Boolean result of the validation
+    """
+    current_timestamp = time.time()
+    is_signature_expired = timestamp < current_timestamp - valid_for
+
+    if is_signature_expired:
+        return False
+
+    payload_hash = compute_hex_hash('{}{}{}'.format(body, timestamp, cloudinary.config().api_secret))
```
Comment (const-cloudinary) [inline cloudinary/utils.py]: Please verify that api_secret is not empty and throw exception if it is. Add unit test.

File: cloudinary/utils.py [hunk 8]
```diff
@@ -1264,6 +1264,56 @@ def wrapper(*args, **kwargs):
...
+    Validates notification signature against Cloudinary configuration
+
+    :param body: Request body str
+    :param timestamp: Request timestamp
+    :param signature: Notification signature
+    :param valid_for: For how long the signature is valid, in seconds
+
+    :return: Boolean result of the validation
+    """
+    current_timestamp = time.time()
```
Comment (const-cloudinary) [inline cloudinary/utils.py]: Why not simplify to:
```
if timestamp < time.time() - valid_for:
    return False
```
?

File: cloudinary/utils.py [hunk 9]
```diff
@@ -1264,6 +1264,56 @@ def wrapper(*args, **kwargs):
...
+    is_signature_expired = timestamp < current_timestamp - valid_for
+
+    if is_signature_expired:
+        return False
+
+    payload_hash = compute_hex_hash('{}{}{}'.format(body, timestamp, cloudinary.config().api_secret))
+
+    return signature == payload_hash
```
Comment (const-cloudinary) [inline cloudinary/utils.py]: `payload_hash` can be omitted, just `return signature == compute_hex_hash(...)`

File: test/test_utils.py [hunk 10]
```diff
@@ -1198,6 +1219,68 @@ def test_build_eager(self):
         for message, value, expected in test_data:
             self.assertEqual(expected, build_eager(value), message)
 
+    def test_compute_hash(self):
+        self.assertEqual("4de279c82056603e91aab3930a593b8b887d9e48",
+                         compute_hex_hash("https://cloudinary.com/images/old_logo.png"))
+
+        original_value = str(uuid.uuid4())
+
+        self.assertEqual(compute_hex_hash(original_value), compute_hex_hash(original_value),
```
Comment (const-cloudinary) [inline test/test_utils.py]: These tests seems a bit redundant (consistency), I would add more tests for edge cases( empty, None, int values, bytes, etc)

File: test/test_utils.py [hunk 11]
```diff
@@ -1198,6 +1219,68 @@ def test_build_eager(self):
...
+
+        self.assertNotEqual(compute_hex_hash(original_value), compute_hex_hash("some string"),
+                            "Unequal inputs hashes should not match")
+
+    def test_verify_api_response_signature(self):
+        public_id = 'tests/logo.png'
+        test_version = 1
+        api_response_signature = '08d3107a5b2ad82e7d82c0b972218fbf20b5b1e0'
+
+        cloudinary.config(api_secret='X7qLTrsES31MzxxkxPPA-pAGGfU')
```
Comment (const-cloudinary) [inline test/test_utils.py]: Please extract `api_secret ` to a constant, it is reused multiple times

File: test/test_utils.py [hunk 12]
```diff
@@ -1198,6 +1219,68 @@ def test_build_eager(self):
...
+
+        self.assertTrue(verify_api_response_signature(public_id, test_version, api_response_signature),
+                        "The response signature is valid for the same parameters")
+
+        self.assertFalse(verify_api_response_signature(public_id, test_version + 1, api_response_signature),
+                         "The response signature is invalid for the wrong version")
+
+        cloudinary.config(api_secret='b')
```
Comment (const-cloudinary) [inline test/test_utils.py]: That is a bad practice, in case assertion fails, we never get to this line, use `with patch()`  for setting `api_key`

File: test/test_utils.py [hunk 13]
```diff
@@ -1198,6 +1219,68 @@ def test_build_eager(self):
...
+
+            self.assertFalse(verify_notification_signature(body, valid_response_timestamp,
+                                                           signature + 'chars'),
+                             "{} invalid for non matching and not expired signature".format(test_message_part))
+
+            self.assertFalse(verify_notification_signature(body, valid_response_timestamp,
+                                                           signature + 'chars',
+                                                           valid_for - 1),
+                             "{} invalid for non matching and expired signature".format(test_message_part))
+        cloudinary.config(api_secret='b')
```
Comment (const-cloudinary) [inline test/test_utils.py]: That is a bad practice, in case assertion fails, we never get to this line, use `with patch()`  for setting `api_key`
